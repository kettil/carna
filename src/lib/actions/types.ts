import { Watermark, Watermarks } from 'istanbul-lib-report';
import { SpawnKillHandler } from '../utils/createSpawnKillHandler';

type BabelActionProps = {
  watch?: boolean;
  spawnKillHandler?: SpawnKillHandler;
};

type BabelNodeActionProps = {
  watch: boolean;
  script: string;
  watchPaths?: string[];
};

type CommitlintActionProps = {
  edit: string;
};

type CoverageWatermarkThreshold = Watermark | number | undefined;
type CoverageActionProps = {
  projects: string[];
  watermarks?: Partial<Record<keyof Watermarks, CoverageWatermarkThreshold>>;
};

type DepcheckActionProps = {
  path: string;
  ignorePackages?: string[];
};

type EslintActionProps = {
  write?: boolean;
  files?: string[];
};

type JestActionProps = {
  projects: string[];
  updateSnapshot?: boolean;
  runInBand?: boolean;
  verbose?: boolean;
  watch?: boolean;
};

type LicensecheckActionProps = {
  ignorePackages: string[];
  path?: string;
};

type PrettierActionProps = {
  write?: boolean;
  extension?: string;
  files?: string[];
};

type SemverActionProps = {
  path?: string;
};

type TscActionProps = {
  mode: 'type-check' | 'type-create';
};

export type {
  BabelActionProps,
  BabelNodeActionProps,
  CommitlintActionProps,
  CoverageActionProps,
  CoverageWatermarkThreshold,
  DepcheckActionProps,
  EslintActionProps,
  JestActionProps,
  LicensecheckActionProps,
  PrettierActionProps,
  SemverActionProps,
  TscActionProps,
};
