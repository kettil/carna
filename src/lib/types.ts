import { Logger } from './cli/logger';

export type NpmInstallMode = 'prod' | 'dev' | 'optional';

export type PropsGlobal = {
  readonly cwd: string;
  readonly tpl: string;
  readonly cfg: string;
  readonly log: Logger;
  readonly vvv: boolean;
};

export type Action<
  Props extends Record<string, unknown> | undefined = undefined,
  ActionReturn = void
> = Props extends undefined
  ? (argv: PropsGlobal) => Promise<ActionReturn>
  : (argv: PropsGlobal, props: Props) => Promise<ActionReturn>;
