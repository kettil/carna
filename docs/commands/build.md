# The `build` task

Build the application

## Options

| Options | Description       |
| ------- | ----------------- |
| --ci    | Run it in CI mode |

## Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
