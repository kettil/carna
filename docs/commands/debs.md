# The `debs` task

Checks if there are orphaned packages

## Options

| Options | Description       |
| ------- | ----------------- |
| --ci    | Run it in CI mode |

## Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
