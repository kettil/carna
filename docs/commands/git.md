# The `git` tasks

The tasks will be automatically called by husky during commit.

## The `git commit` task

Checks the files in staged with type, prettier and eslint.

### Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |

## The `git msg` task

Checks the git commit message for a consistent structure.

### Options

| Options | Description                     |
| ------- | ------------------------------- |
| --edit  | Path to the COMMIT_EDITMSG file |

### Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |

## The `git push` task

Checks the files with type, prettier and eslint.

### Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
