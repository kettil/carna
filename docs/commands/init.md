# The `init` task

Creates a new project with all config files in the current folder.

## Options

| Options    | short | Description                                                  |
| ---------- | ----- | ------------------------------------------------------------ |
| --package  | -p    | Project is created as a package                              |
| --cli      | -c    | Extends the package with CLI features (required `--package`) |
| --noCommit |       | No initial commit is executed at the end                     |

## Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
