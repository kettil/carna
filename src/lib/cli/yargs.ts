import { CommandModule, Arguments, Argv } from 'yargs';
import ExecutableError from '../errors/executableError';
import { PropsGlobal } from '../types';

export type CommandModuleCommand = string;

export type CommandModuleDescribe = NonNullable<CommandModule<PropsGlobal>['describe']>;

export type CommandModuleBuilder<Props = Record<string, unknown>> = (
  yargs: Argv<PropsGlobal>,
) => Argv<PropsGlobal & Props>;

export type CommandModuleHandler<Props = Record<string, unknown>> = (args: Arguments<PropsGlobal & Props>) => void;

export const builderDefault = <Props>(cmd: string, fn: CommandModuleBuilder<Props>): CommandModuleBuilder<Props> => (
  yargs,
) => fn(yargs).usage(`Usage: $0 ${cmd} [options]`).help().version(false)
  .showHelpOnFail(false);

export const errorHandler = (argv: PropsGlobal, error: unknown): void => {
  if (error instanceof ExecutableError) {
    argv.log.log(error.stdout);

    /* eslint-disable-next-line node/no-process-exit, unicorn/no-process-exit */
    process.exit(1);
  }

  throw error;
};
