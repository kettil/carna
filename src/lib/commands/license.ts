import { createBuilder, createHandler } from '../cli/yargs';
import licenseTask, { LicenseProps } from '../tasks/licenseTask';

export const command = 'license';
export const desc = 'Run the license check';

export const handler = createHandler<LicenseProps>(licenseTask);
export const builder = createBuilder<LicenseProps>(command, (yargs) => yargs);
