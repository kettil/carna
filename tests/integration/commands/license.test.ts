import * as fs from 'fs';
import { Argv } from 'yargs';
import { handler, builder, command, desc } from '../../../src/lib/commands/license';
import { PropsGlobal } from '../../../src/lib/types';
import { getArgv } from '../../shared/setup/argv';
import { getReadFileWithHooksFiles } from '../../shared/setup/readFileFiles';

describe('command license', () => {
  test('it should be complete the yargs command structure', () => {
    expect(typeof command).toBe('string');
    expect(typeof desc).toBe('string');
    expect(typeof builder).toBe('function');
    expect(typeof handler).toBe('function');
  });

  test('it should be create the expected builder', () => {
    expect.assertions(1);

    const yargs = {
      options: jest.fn((...args: unknown[]) => {
        expect(args).toMatchSnapshot('options');

        return yargs;
      }),
      usage: jest.fn((...args: unknown[]) => {
        expect(args).toMatchSnapshot('usage');

        return yargs;
      }),

      help: jest.fn(() => yargs),
      version: jest.fn(() => yargs),
      showHelpOnFail: jest.fn(() => yargs),
    } as unknown as Argv<PropsGlobal>;

    builder(yargs);
  });

  test('it should work', async () => {
    await handler(getArgv({ hook: 'msg', edit: 'path/to/commit.message' }));

    expect(true).toBeTruthy();
  });

  test('it should work with hooks', async () => {
    (fs as any).setMockReadFileFiles(getReadFileWithHooksFiles());

    await handler(getArgv({ hook: 'msg', edit: 'path/to/commit.message' }));

    expect(true).toBeTruthy();
  });
});
