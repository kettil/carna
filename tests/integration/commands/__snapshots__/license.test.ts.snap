// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command license it should be create the expected builder: usage 1`] = `
Array [
  "Usage: $0 license [options]",
]
`;

exports[`command license it should work with hooks: child-process 1`] = `
Object {
  "args": Array [
    "run",
    "precarna:license",
  ],
  "command": "npm",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command license it should work with hooks: child-process 2`] = `
Object {
  "args": Array [
    "run",
    "postcarna:license",
  ],
  "command": "npm",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;
